experiments/2018-05-23-rust-on-android-JNA.md

```diff
@@ -0,0 +1,205 @@
+
+# Building our library
+
+Now we're all set up and we're ready to start. Let's create the `lib` directory. If you've already created a Rust project from following either or our previous posts, you don't need to do it again.
+
+
+First create a directory for our project.
+
+```
+mkdir greetings
+cd greetings
+cargo new rust --lib
+mkdir android
+```
+
+`cargo new rust` sets up a brand new Rust project with its default files and directories in a directory called `rust`. The name of the directory is not important. In this directory is a file called `Cargo.toml`, which is the package manager descriptor file, and there is a subdirectory, `src`, which contains a file called `lib.rs`. This will contain the Rust code that we will be executing.
+
+Our Rust project here is a super simple Hello World library. It contains a function `rust_greeting` that takes a string argument and returns a greeting including that argument. Therefore, if the argument is "world", the returned string is "Hello world".
+
+Open `cargo/src/lib.rs` and enter the following code.
+
+```rust
+use std::os::raw::{c_char};
+use std::ffi::{CString, CStr};
+
+#[no_mangle]
+pub extern "C" fn rust_greeting(to: *const c_char) -> *mut c_char {
+    let c_str = unsafe { CStr::from_ptr(to) };
+    let recipient = match c_str.to_str() {
+        Err(_) => "there",
+        Ok(string) => string,
+    };
+
+    CString::new("Hello ".to_owned() + recipient).unwrap().into_raw()
+}
+```
+
+Let's take a look at what is going on here.
+
+As we will be calling this library from non-Rust code, we will actually be calling it through a C bridge. `#[no_mangle]` tells the compiler not to mangle the function name as it usually does by default, ensuring our function name is exported as if it had been written in C.
+
+`extern` tells the Rust compiler that this function will be called from outside of Rust and to therefore ensure that it is compiled using C calling conventions.
+
+The string that `rust_greeting` accepts is a pointer to a C char array. We have to then convert the string from a C string to a Rust `str`. First we create a `CStr` object from the pointer. We then convert it to a `str` and check the result. If an error has occurred, then no arg was provided and we substitute `there`, otherwise we use the value of the provided string. We then append the provided string on the end of our greeting string to create our return string. The return string is then converted into a `CString` and passed back into C code.
+
+In order to build our library, we need to tell the compiler what type of a library it should produce. You can specify this in the `Cargo.toml` file's `[lib]` section:
+
+```
+[lib]
+crate-type = ["dylib"]
```

Comment:
@fluffyemily we had to use `cdylib` in our fxa-rust-client. In projects that have `*.so` dependencies of dependencies with `dylib` you would get `libstd-[hex_value].so`(Rust stdlib?) not found. 

In addition the generated size of `cdylib` is smaller.

Ref: https://stackoverflow.com/a/44175499/216672

Reply:
I'm not sure what is the best way to explain the differences and usages of the above in this blog post.